### PR TITLE
[Node] Utility methods for ObjectPathPair handling

### PR DESCRIPTION
**Pull request by Lunderberg**
This commit adds a templated overload to `SEqualReducer::operator()` that accepts a lambda function to update the path of the LHS and RHS of the comparison.

```c++
// Usage prior to this utility function
if (equal.IsPathTracingEnabled()) {
  const ObjectPathPair& self_paths = equal.GetCurrentObjectPaths();
  ObjectPathPair attr_paths = {self_paths->lhs_path->Attr("value"),
                               self_paths->rhs_path->Attr("value")};
  if (!equal(this->value, other->value, attr_paths)) return false;
} else {
  if (!equal(this->value, other->value)) return false;
}

// Usage after this utility function
if (!equal(this->value, other->value,
           [](const auto& path) { return path->Attr("value"); })) {
  return false;
}
```

### CODE REVIEW

**Comment by Lunderberg** [comment]
This arose during debugging for https://github.com/apache/tvm/pull/14486, but wasn't strictly necessary for it.

File: src/ir/module.cc [hunk 1]
```diff
@@ -63,46 +63,46 @@ IRModule::IRModule(tvm::Map<GlobalVar, BaseFunc> functions,
...
   }
 
-  if (type_definitions.size() != other->type_definitions.size()) return false;
-  // Update GlobalTypeVar remap
-  for (const auto& gtv : this->GetGlobalTypeVars()) {
-    if (!other->ContainGlobalTypeVar(gtv->name_hint)) return false;
-    if (!equal.DefEqual(gtv, other->GetGlobalTypeVar(gtv->name_hint))) return false;
+  // Checking functions and type definitions
+  if (!equal(this->functions, other->functions,
+             [](const auto& path) { return path->Attr("functions"); })) {
```
Comment (vinx13) [inline src/ir/module.cc:98]: Is `func_path` (adding function name to the path) handled like before?

Comment (Lunderberg) [reply src/ir/module.cc:98]: It is, yes.  The `IRModule`'s handler only needs to handle appending `.functions` to the path, with the function name added by the equality check on `Map`.  This results in a path of `<root>.functions[I.GlobalVar("func")].body...`.

I've added a unit test that verifies that the error message includes the function name, as that's definitely a good thing to verify for user-friendliness.

**Comment by Lunderberg** [comment]
@tvm-bot rerun

**Review by vinx13 — APPROVED** [review]

**Pull request merged**